File: src/install/isolated_install/Hardlinker.zig
const Hardlinker = @This();

src_dir: FD,
src: bun.AbsPath(.{ .sep = .auto, .unit = .os }),
dest: bun.Path(.{ .sep = .auto, .unit = .os }),
walker: Walker,

pub fn init(
    folder_dir: FD,
    src: bun.AbsPath(.{ .sep = .auto, .unit = .os }),
    dest: bun.Path(.{ .sep = .auto, .unit = .os }),
    skip_dirnames: []const bun.OSPathSlice,
) OOM!Hardlinker {
    return .{
        .src_dir = folder_dir,
        .src = src,
        .dest = dest,
        .walker = walker: {
            var w = try Walker.walk(
                folder_dir,
                bun.default_allocator,
                &.{},
                skip_dirnames,
            );
            w.resolve_unknown_entry_types = true;
            break :walker w;
        },
    };
}

pub fn deinit(this: *Hardlinker) void {
    this.walker.deinit();
}

pub fn link(this: *Hardlinker) OOM!sys.Maybe(void) {
    if (bun.install.PackageManager.verbose_install) {
        bun.Output.prettyErrorln(
            \\Hardlinking {f} to {f}
        ,
            .{
                bun.fmt.fmtOSPath(this.src.slice(), .{ .path_sep = .auto }),
                bun.fmt.fmtOSPath(this.dest.slice(), .{ .path_sep = .auto }),
            },
        );
        bun.Output.flush();
    }

    if (comptime Environment.isWindows) {
        const cwd_buf = bun.w_path_buffer_pool.get();
        defer bun.w_path_buffer_pool.put(cwd_buf);
        const dest_cwd = FD.cwd().getFdPathW(cwd_buf) catch {
            return .initErr(bun.sys.Error.fromCode(bun.sys.E.ACCES, .link));
        };

        while (switch (this.walker.next()) {
            .result => |res| res,
            .err => |err| return .initErr(err),
        }) |entry| {
            var src_save = this.src.save();
            defer src_save.restore();

            this.src.append(entry.path);

            var dest_save = this.dest.save();
            defer dest_save.restore();

            this.dest.append(entry.path);

            switch (entry.kind) {
                .directory => {
                    FD.cwd().makePath(u16, this.dest.slice()) catch {};
                },
                .file => {
                    const destfile_path_buf = bun.w_path_buffer_pool.get();
                    const destfile_path_buf2 = bun.w_path_buffer_pool.get();
                    defer bun.w_path_buffer_pool.put(destfile_path_buf2);
                    defer bun.w_path_buffer_pool.put(destfile_path_buf);
                    // `dest` may already be absolute (global virtual store
                    // entries live under the cache, not cwd); only prefix the
                    // working-directory path when it's project-relative.
                    const dest_parts: []const []const u16 = if (this.dest.len() > 0 and bun.path.Platform.windows.isAbsoluteT(u16, this.dest.slice()))
                        &.{this.dest.slice()}
                    else
                        &.{ dest_cwd, this.dest.slice() };
                    const destfile_path = bun.strings.addNTPathPrefixIfNeeded(destfile_path_buf2, bun.path.joinStringBufWZ(destfile_path_buf, dest_parts, .windows));

                    const srcfile_path_buf = bun.w_path_buffer_pool.get();
                    defer bun.w_path_buffer_pool.put(srcfile_path_buf);

                    switch (sys.link(u16, this.src.sliceZ(), destfile_path)) {
                        .result => {},
                        .err => |link_err1| switch (link_err1.getErrno()) {
                            .UV_EEXIST,
                            .EXIST,
                            => {
                                if (bun.install.PackageManager.verbose_install) {
                                    bun.Output.prettyErrorln(
                                        \\Hardlinking {f} to a path that already exists: {f}
                                    ,
                                        .{
                                            bun.fmt.fmtOSPath(this.src.slice(), .{ .path_sep = .auto }),
                                            bun.fmt.fmtOSPath(destfile_path, .{ .path_sep = .auto }),
                                        },
                                    );
                                }

                                try_delete: {
                                    const delete_tree_buf = bun.path_buffer_pool.get();
                                    defer bun.path_buffer_pool.put(delete_tree_buf);

                                    const delete_tree_path = bun.strings.convertUTF16toUTF8InBuffer(delete_tree_buf, this.dest.slice()) catch {
                                        break :try_delete;
                                    };
                                    FD.cwd().deleteTree(delete_tree_path) catch {};
                                }
                                switch (sys.link(u16, this.src.sliceZ(), destfile_path)) {
                                    .result => {},
                                    .err => |link_err2| switch (link_err2.getErrno()) {
                                        .UV_EEXIST, .EXIST => {},
                                        else => return .initErr(link_err2),
                                    },
                                }
                            },
                            .UV_ENOENT,
                            .NOENT,
                            => {
                                if (bun.install.PackageManager.verbose_install) {
                                    bun.Output.prettyErrorln(
                                        \\Hardlinking {f} to a path that doesn't exist: {f}
                                    ,
                                        .{
                                            bun.fmt.fmtOSPath(this.src.slice(), .{ .path_sep = .auto }),
                                            bun.fmt.fmtOSPath(destfile_path, .{ .path_sep = .auto }),
                                        },
                                    );
                                }
                                const dest_parent = this.dest.dirname() orelse {
                                    return .initErr(link_err1);
                                };

                                FD.cwd().makePath(u16, dest_parent) catch {};

                                switch (sys.link(u16, this.src.sliceZ(), destfile_path)) {
                                    .result => {},
                                    .err => |link_err2| switch (link_err2.getErrno()) {
                                        .UV_EEXIST, .EXIST => {},
                                        else => return .initErr(link_err2),
                                    },
                                }
                            },
                            else => return .initErr(link_err1),
                        },
                    }
                },
                else => {},
            }
        }

        return .success;
    }

    while (switch (this.walker.next()) {
        .result => |res| res,
        .err => |err| return .initErr(err),
    }) |entry| {
        var dest_save = this.dest.save();
        defer dest_save.restore();

        this.dest.append(entry.path);

        switch (entry.kind) {
            .directory => {
                FD.cwd().makePath(u8, this.dest.sliceZ()) catch {};
            },
            .file => {
                switch (sys.linkatZ(entry.dir, entry.basename, FD.cwd(), this.dest.sliceZ())) {
                    .result => {},
                    .err => |link_err1| {
                        switch (link_err1.getErrno()) {
                            .EXIST => {
                                FD.cwd().deleteTree(this.dest.slice()) catch {};
                                switch (sys.linkatZ(entry.dir, entry.basename, FD.cwd(), this.dest.sliceZ())) {
                                    .result => {},
                                    // Another writer (concurrent install into a
                                    // shared global-store entry) re-created the
                                    // same hardlink between our delete and this
                                    // retry; it points at identical bytes.
                                    .err => |link_err2| if (link_err2.getErrno() != .EXIST) return .initErr(link_err2),
                                }
                            },
                            .NOENT => {
                                const dest_parent = this.dest.dirname() orelse {
                                    return .initErr(link_err1);
                                };

                                FD.cwd().makePath(u8, dest_parent) catch {};
                                switch (sys.linkatZ(entry.dir, entry.basename, FD.cwd(), this.dest.sliceZ())) {
                                    .result => {},
                                    .err => |link_err2| if (link_err2.getErrno() != .EXIST) return .initErr(link_err2),
                                }
                            },
                            else => return .initErr(link_err1),
                        }
                    },
                }
            },
            else => {},
        }
    }

    return .success;
}

const Walker = @import("../../walker_skippable.zig");

const bun = @import("bun");
const Environment = bun.Environment;
const FD = bun.FD;
const OOM = bun.OOM;
const sys = bun.sys;

File: src/install/isolated_install/Hardlinker.rs
use bun_alloc::AllocError;
#[cfg(not(windows))]
use bun_sys::FdDirExt;
use bun_sys::walker_skippable::Walker;
use bun_sys::{self as sys, EntryKind, Fd, FdExt};
// OS-unit paths are u8 on POSIX, u16
// on Windows — encoded here via the `OSPathChar` type alias so the struct's
// `slice()`/`slice_z()` produce the platform-native width without per-field
// `#[cfg]` divergence.
#[cfg(windows)]
use bun_paths::path_options::AssumeOk as _;
use bun_paths::{AbsPath, OSPathChar, OSPathSlice, Path};

type OsAbsPath = AbsPath<OSPathChar, { bun_paths::path_options::PathSeparators::AUTO }>;
type OsPath = Path<
    OSPathChar,
    { bun_paths::path_options::Kind::ANY },
    { bun_paths::path_options::PathSeparators::AUTO },
>;

pub(crate) struct Hardlinker {
    pub(crate) src: OsAbsPath,
    pub(crate) dest: OsPath,
    pub(crate) walker: Walker,
}

impl Hardlinker {
    pub(crate) fn init(
        folder_dir: Fd,
        src: OsAbsPath,
        dest: OsPath,
        skip_dirnames: &[&OSPathSlice],
    ) -> Result<Hardlinker, AllocError> {
        Ok(Hardlinker {
            src,
            dest,
            walker: {
                let mut w = bun_sys::walker_skippable::walk(
                    folder_dir,
                    // bun.default_allocator dropped — global mimalloc
                    &[],
                    skip_dirnames,
                )?;
                w.resolve_unknown_entry_types = true;
                w
            },
        })
    }

    pub(crate) fn link(&mut self) -> Result<sys::Result<()>, AllocError> {
        if crate::PackageManager::verbose_install() {
            bun_core::pretty_errorln!(
                "Hardlinking {} to {}",
                bun_core::fmt::fmt_os_path(self.src.slice(), Default::default()),
                bun_core::fmt::fmt_os_path(self.dest.slice(), Default::default()),
            );
            bun_core::output::flush();
        }

        #[cfg(windows)]
        {
            let mut cwd_buf = bun_paths::w_path_buffer_pool::get();
            let dest_cwd_len = bun_core::strings::convert_utf8_to_utf16_in_buffer(
                &mut cwd_buf[..],
                crate::bun_fs::FileSystem::instance().top_level_dir(),
            )
            .len();

            loop {
                let entry = match self.walker.next() {
                    sys::Result::Ok(Some(res)) => res,
                    sys::Result::Ok(None) => break,
                    sys::Result::Err(err) => return Ok(sys::Result::Err(err)),
                };

                // A `path.save()` ResetScope would hold `&mut Path` and keep
                // `self.src`/`self.dest` exclusively borrowed for the rest of
                // the iteration. Capture the saved length directly and restore
                // via `set_length` after the body (and before any error return)
                // so the truncation happens on every exit.
                let src_saved_len = self.src.len();
                // `OsAbsPath`/`OsPath` use `CheckLength::ASSUME`, so `append`'s
                // `Err(MaxPathExceeded)` arm is statically unreachable -- see
                // `path_options::AssumeOk`.
                self.src.append(entry.path.as_slice()).assume_ok();

                let dest_saved_len = self.dest.len();
                self.dest.append(entry.path.as_slice()).assume_ok();

                let err: Option<sys::Error> = 'body: {
                    match entry.kind {
                        EntryKind::Directory => {
                            let _ = sys::make_path::make_path::<u16>(
                                &sys::Dir::cwd(),
                                self.dest.slice(),
                            );
                        }
                        EntryKind::File => {
                            let mut destfile_path_buf = bun_paths::w_path_buffer_pool::get();
                            let mut destfile_path_buf2 = bun_paths::w_path_buffer_pool::get();
                            // `dest` may already be absolute (global virtual store
                            // entries live under the cache, not cwd); only prefix the
                            // working-directory path when it's project-relative.
                            // Stash the dest slice once so the `&self` borrow
                            // doesn't span the buffer-mut below.
                            let dest_slice: &[u16] = self.dest.slice();
                            let dest_parts: &[&[u16]] = if !dest_slice.is_empty()
                                && bun_paths::Platform::Windows.is_absolute_t::<u16>(dest_slice)
                            {
                                &[dest_slice]
                            } else {
                                &[&cwd_buf[..dest_cwd_len], dest_slice]
                            };
                            let joined = bun_paths::resolve_path::join_string_buf_w_same::<
                                bun_paths::platform::Windows,
                            >(
                                &mut destfile_path_buf[..], dest_parts
                            );
                            let destfile_path = bun_paths::strings::add_nt_path_prefix_if_needed(
                                &mut destfile_path_buf2[..],
                                joined,
                            );

                            match sys::link_w(self.src.slice_z(), destfile_path) {
                                sys::Result::Ok(()) => {}
                                sys::Result::Err(link_err1) => match link_err1.get_errno() {
                                    sys::E::UV_EEXIST | sys::E::EEXIST => {
                                        if crate::PackageManager::verbose_install() {
                                            bun_core::pretty_errorln!(
                                                "Hardlinking {} to a path that already exists: {}",
                                                bun_core::fmt::fmt_os_path(
                                                    self.src.slice(),
                                                    Default::default()
                                                ),
                                                bun_core::fmt::fmt_os_path(
                                                    destfile_path.as_slice(),
                                                    Default::default()
                                                ),
                                            );
                                        }

                                        {
                                            let mut delete_tree_buf =
                                                bun_paths::path_buffer_pool::get();

                                            let delete_tree_path =
                                                bun_core::convert_utf16_to_utf8_in_buffer(
                                                    &mut delete_tree_buf[..],
                                                    self.dest.slice(),
                                                );
                                            let _ = Fd::cwd().delete_tree(delete_tree_path);
                                        }
                                        match sys::link_w(self.src.slice_z(), destfile_path) {
                                            sys::Result::Ok(()) => {}
                                            sys::Result::Err(link_err2) => {
                                                break 'body Some(link_err2);
                                            }
                                        }
                                    }
                                    sys::E::UV_ENOENT | sys::E::ENOENT => {
                                        if crate::PackageManager::verbose_install() {
                                            bun_core::pretty_errorln!(
                                                "Hardlinking {} to a path that doesn't exist: {}",
                                                bun_core::fmt::fmt_os_path(
                                                    self.src.slice(),
                                                    Default::default()
                                                ),
                                                bun_core::fmt::fmt_os_path(
                                                    destfile_path.as_slice(),
                                                    Default::default()
                                                ),
                                            );
                                        }
                                        let Some(dest_parent) = self.dest.dirname() else {
                                            break 'body Some(link_err1);
                                        };

                                        let _ = sys::make_path::make_path::<u16>(
                                            &sys::Dir::cwd(),
                                            dest_parent,
                                        );

                                        match sys::link_w(self.src.slice_z(), destfile_path) {
                                            sys::Result::Ok(()) => {}
                                            sys::Result::Err(link_err2) => {
                                                break 'body Some(link_err2);
                                            }
                                        }
                                    }
                                    _ => break 'body Some(link_err1),
                                },
                            }
                        }
                        _ => {}
                    }
                    None
                };

                self.src.set_length(src_saved_len);
                self.dest.set_length(dest_saved_len);

                if let Some(err) = err {
                    return Ok(sys::Result::Err(err));
                }
            }

            return Ok(sys::Result::Ok(()));
        }

        #[cfg(not(windows))]
        {
            loop {
                let entry = match self.walker.next() {
                    sys::Result::Ok(Some(res)) => res,
                    sys::Result::Ok(None) => break,
                    sys::Result::Err(err) => return Ok(sys::Result::Err(err)),
                };

                // A `dest.save()` ResetScope would hold `&mut Path` and keep
                // `self.dest` exclusively borrowed across the body. Capture
                // `len()` and restore via `set_length()` after the body so the
                // truncation runs on every exit.
                let dest_saved_len = self.dest.len();
                let _ = self.dest.append(entry.path.as_bytes()); // OOM/capacity: fire-and-forget

                let err: Option<sys::Error> = 'body: {
                    match entry.kind {
                        EntryKind::Directory => {
                            let _ = Fd::cwd().make_path(self.dest.slice());
                        }
                        EntryKind::File => {
                            match sys::linkat(
                                entry.dir,
                                entry.basename,
                                Fd::cwd(),
                                self.dest.slice_z(),
                            ) {
                                sys::Result::Ok(()) => {}
                                sys::Result::Err(link_err1) => match link_err1.get_errno() {
                                    sys::E::EEXIST => {
                                        let _ = Fd::cwd().delete_tree(self.dest.slice());
                                        match sys::linkat(
                                            entry.dir,
                                            entry.basename,
                                            Fd::cwd(),
                                            self.dest.slice_z(),
                                        ) {
                                            sys::Result::Ok(()) => {}
                                            sys::Result::Err(link_err2) => {
                                                break 'body Some(link_err2);
                                            }
                                        }
                                    }
                                    sys::E::ENOENT => {
                                        let Some(dest_parent) = self.dest.dirname() else {
                                            break 'body Some(link_err1);
                                        };

                                        let _ = Fd::cwd().make_path(dest_parent);
                                        match sys::linkat(
                                            entry.dir,
                                            entry.basename,
                                            Fd::cwd(),
                                            self.dest.slice_z(),
                                        ) {
                                            sys::Result::Ok(()) => {}
                                            sys::Result::Err(link_err2) => {
                                                break 'body Some(link_err2);
                                            }
                                        }
                                    }
                                    _ => break 'body Some(link_err1),
                                },
                            }
                        }
                        _ => {}
                    }
                    None
                };

                self.dest.set_length(dest_saved_len);

                if let Some(err) = err {
                    return Ok(sys::Result::Err(err));
                }
            }

            Ok(sys::Result::Ok(()))
        }
    }
}
